dbms/src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -2882,7 +2882,14 @@ void StorageReplicatedMergeTree::shutdown()
 
     if (queue_task_handle)
         global_context.getBackgroundPool().removeTask(queue_task_handle);
-    queue_task_handle.reset();
+
+    {
+        /// Queue can trigger queue_task_handle itself. So we ensure that all
```

Comment:
Maybe it's also possible with atomic flag?

Reply:
But how? Operations with the flag are atomic, but `t1` can check the flag and it will say that pointer is OK then sleep for some time after that `shutdown` thread `t2` will be called and sets both flag and pointer to zero. After that `t1` will wake up and call `wake` on empty pointer.

Reply:
You're right.